lineapy/graph_reader/session_artifacts.py

```diff
@@ -264,24 +264,40 @@ def _update_node_context(self):
                         len(self.node_context[node_id].dependent_variables)
                         == 0
                     )
-                    has_assigned_before = (
-                        self.input_parameters_node.get(var, None) is not None
+                    previous_assignment = self.input_parameters_node.get(
+                        var, None
                     )
-                    if is_literal_assignment:
-                        if has_assigned_before:
+
+                    if previous_assignment is None:
+                        self.input_parameters_node[var] = node_id
+                    else:
+                        previous_assignment_is_literal = (
+                            len(
+                                self.node_context[
+                                    previous_assignment
+                                ].dependent_variables
+                            )
+                            == 0
+                        )
+
+                        if (
+                            is_literal_assignment
+                            and previous_assignment_is_literal
+                        ):
                             raise ValueError(
-                                "Variable %s, is defined more than once", var
+                                f"Variable {var}, is defined more than once"
                             )
-                        else:
+                        elif not previous_assignment_is_literal:
+                            # previous assignment is not literal, so we can reassign it
                             self.input_parameters_node[var] = node_id
+                        # else previous assignment is literal and we should not override it
 
         for var, node_id in self.input_parameters_node.items():
             if len(self.node_context[node_id].dependent_variables) > 0:
-                # Duplicated variable name existing
-                logger.error(
-                    "LineaPy only supports literal value as input parameters for now."
+                raise ValueError(
+                    f"LineaPy only supports literal value as input parameters for now. "
+                    f"{var} only has non-literal values in this Session."
```

Comment:
This line sounds odd. If var had some literal value and reassigned with some non-literal value, would we automatically assume the literal value is what the user meant?

Reply:
Yeah ... the code by default would assign to the literal value. This might be a bit non-intuitive ... we can change it to error with duplicate value error here. wdyt? @mingjerli

Reply:
Decision to keep behavior as is, meaning this error message should stand as written. Added a test case to cover as well.